src/test/java/org/jooq/lambda/SeqTest.java

```diff
@@ -1543,6 +1543,22 @@ public void testIterate() {
     }
 
     @Test
+    public void testIterateWhilePresent() {
+        assertEquals(
+              asList(),
+              Seq.iterateWhilePresent(null, i -> Optional.empty()).toList()
+        );
```

Comment:
In this case, it's a return type, not a field/parameter type, so it works along the ideas of the designers, in a way. Even better than Kotlin: Ceylon, with real union types. Your function would return a type like `T|EndOfIteration`

Which reminds me that Java's checked exceptions are really union types (if you squint really hard). Perhaps, we could abort iteration with a checked exception? ;-) (semi-serious)